jacovid/featurize.py

```diff
@@ -0,0 +1,151 @@
+def featurize_data(data, target_var="inc_hosp", h=1, feature_parameters = 
+    [{
+        "fun": 'moving_average',
+        "args": {'target_var': "inc_hosp", 'num_days': 7}
+        },
+        {
+        "fun": 'lagged_values',
+        "args": {'target_var': "inc_hosp", 'window_size': 2}
+        },
+    ]):
+    """
+    Convert data to tensors containing features x and responses y for each location
+    and time.
+
+    Parameters
+    ----------
+    data: data frame 
+        It has columns location, date, and a column with the response variable to forecast.
+        This data frame needs to be sorted by location and date columns in ascending order.
+    target_var: string
+        Name of the column in the data frame with the forecast target variable.
+        Default to "inc_hosp"
+    h: integer
+        Forecast horizon. Default to 1.
+    feature_parameters: array of dictionaries
+        List of features to calculate. Each dictionary should have a `fun` key with feature function name
+        and an `args` key with parameter name and values of this function.
+        Available feature functions are "moving_average" and "lagged_values".
+        
+    Returns
+    -------
+    x_train_val: 3D tensor with shape (L, T, P) 
+        L is the number of location l and T is the number of time point t for which
+        the full feature vector x_{l,t}, possibly including lagged covariate values,
+        and the response y_{l,t}, corresponding to the target variable at time t+h,
+        could be calculated. P is number of features.
+        Each row is a vector x_{l,t} = [x_{l,t,1},...,x_{l,t,P}] of features for some pair 
+        (l, t) in the training set.
+    y_train_val: 2D tensor with with shape (L, T) 
+        Each value is a forecast target variable value in the training set.
+        y_{l, t} = z_{l, 1, t+h}
+    x_T: 2D tensor with shape (L, P)
+        Each value is test set feature for each location at forecast date.
+    """
+    assert target_var in data.columns
+
+    # data has columns date, location, inc hosp, rate and population
+    T = max(data['date'])
+
+    # calculate features based on given parameters
+    # and collect a list of feature names
+    features_list = list()
+    for feature in feature_parameters:
+        data, features_list = data.pipe(eval(feature["fun"]), features_list =features_list, **feature["args"])
+    
+    # create a column for h horizon ahead target for observed values. 
+    # for each location, this column has h nans in the end.
+    # the last nan is for forecast date.
+    data['h_days_ahead_target'] = data.groupby('location')[target_var].shift(-h)
+
+    # create x_T using data with date = forecast_date (T)
+    data_T = data.loc[data["date"]== T,:]
+
+    # x_T is (L, P)
+    x_T = data_T[features_list].values
+    
+    # list of indices of rows that have as least one nan
+    na_idx, _ = np.where(data.isna())
+    
+    # take out nans in data
+    train_val = data.drop(na_idx)
```

Comment:
could we just use `data.dropna()`?
```suggestion
    # take out nans in data
    train_val = data.dropna()
```